src/rpc/services/JoinChannelService.ts

```diff
@@ -21,68 +37,316 @@ class JoinChannelService extends ARpcService {
+            }
+
+            // Process confirmation signature if present
+            if (confirmationSignature) {
+                // Validate the confirmation signature
+                const validationResult =
+                    await this.validateConfirmationSignature(
+                        joinChannel,
+                        confirmationSignature
+                    );
+
+                if (validationResult !== ValidationFlag.VALID) {
+                    console.warn(
+                        `Invalid confirmation signature: ${ValidationFlag[validationResult]}`
+                    );
+                    return;
+                }
+
+                const confirmerAddress =
+                    EvmUtils.retrieveSignerAddressJoinChannel(
+                        joinChannel,
+                        confirmationSignature
+                    );
+
+                // Store the confirmation signature
+                this.joinChannelMap.tryInsert(key, {
+                    signerAddress: confirmerAddress,
+                    signature: confirmationSignature
+                });
+
+                // Broadcast the incoming signature too
+                this.mainRpcService.rpcProxy
+                    .onJoinChannelRequest(
+                        signedJoinChannel,
+                        confirmationSignature
+                    )
+                    .broadcast();
+            }
+
+            // Add our signature if we haven't already
+            const myAddress =
+                await this.mainRpcService.p2pManager.p2pSigner.getAddress();
+            if (!this.joinChannelMap.hasSignature(key, myAddress)) {
+                // Sign it ourselves
+                const mySignedJC = await EvmUtils.signJoinChannel(
+                    joinChannel,
                     this.mainRpcService.p2pManager.p2pSigner
                 );
+
+                // Add our signature
                 this.joinChannelMap.tryInsert(key, {
-                    signerAddress: jc.participant.toString(),
+                    signerAddress: myAddress,
                     signature: mySignedJC.signature as SignatureLike
                 });
+
+                // Broadcast with our signature
                 this.mainRpcService.rpcProxy
                     .onJoinChannelRequest(
                         signedJoinChannel,
                         mySignedJC.signature as SignatureLike
                     )
                     .broadcast();
-                setTimeout(() => {
-                    this.joinChannelMap.delete(key); // prune expired
-                }, timeRemaining * 1000);
-                return;
             }
-            //confirmation logic
-            if (!confirmationSignature) return;
-            let retrivedAddress = EvmUtils.retrieveSignerAddressJoinChannel(
-                jc,
+
+            // Check if we have all required signatures
+            const activeParticipantsSet = await this.getActiveParticipants(
```

Comment:
can use the latest on chain snapshot particpiants

Reply:
to be continued
#53